### ISSUE TITLE
Potential blocking calls not wrapped in blocking

### ISSUE THREAD

**Issue by froth** [ovotech/meters4s#16]
I don't have hard proof for this, but from my understanding several calls to micrometer contain locks and should be wrapped in `blocking` instead of `delay`.

Example:
https://github.com/ovotech/meters4s/blob/master/core/src/main/scala/com/ovoenergy/meters4s/Reporter.scala#L312

This `delay` calls after some indirection the following method in micrometer:
https://github.com/micrometer-metrics/micrometer/blob/54f89f40743937a5a78fb57eb0ddb6b2e6238be4/micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java#L598

This contains a synchronized block over `meterMapLock` and can block the current thread. Therefore I would argue that this could lead to blocking one of the io-compute threads if several calls hit the lock in parallel.

**Comment by keirlawson** [comment]
That's a great point, thanks for reporting.  Would you be prepared to author a PR to fix this?

**Comment by froth** [comment]
Sure, can do so. Most likely at some point next week, this week is kind of a mess ;)

**Comment by froth** [comment]
I opened a pull request. I tried to be as thorough as I could in searching for other locks in the micrometer implementation. But there were all the clear blocking cases I found.

The only other thing I found is that `AtomicInteger` (used in gauge) is not strictly guaranteed to be non blocking. This depends on the underlying processor architecture. However on x86 / x86_64 it behaves in a non blocking way. I did a quick search for arm but did not find clear information. I decided to leave `AtomicInteger` alone for now(my code runs on x86_64 anyways). But maybe someone else wants to take a deeper look.
